File: lib/domain/agent/react_strategy.dart
import 'dart:async';
import 'dart:convert';

import 'package:aios/domain/agent/agent_strategy.dart';
import 'package:aios/domain/agent/agent_tool.dart';
import 'package:aios/domain/agent/audit_log.dart';
import 'package:aios/domain/agent/confirmation_gate.dart';
import 'package:aios/domain/agent/conversation_context.dart';
import 'package:aios/domain/agent/error_recovery.dart';
import 'package:aios/domain/agent/extended_tool.dart';
import 'package:aios/domain/agent/llm_engine.dart';
import 'package:aios/domain/agent/loop_detector.dart';
import 'package:aios/domain/agent/permission_gate.dart';
import 'package:aios/domain/agent/risk_classifier.dart';
import 'package:aios/domain/agent/tool_arg_inference.dart';
import 'package:aios/domain/agent/tool_context.dart';
import 'package:aios/domain/agent/tool_json_parser.dart';
import 'package:aios/domain/agent/tool_permission_mapper.dart';
import 'package:aios/domain/agent/tool_preference_tracker.dart';
import 'package:aios/domain/agent/tool_result.dart';
import 'package:aios/domain/agent/truncate.dart';
import 'package:aios/domain/entities/agent_models.dart';

class ReactStrategy implements AgentStrategy {
  ReactStrategy({
    required LlmEngine engine,
    ToolContext? toolContext,
    Map<String, AgentTool>? basicTools,
    Map<String, ExtendedTool>? extendedTools,
  }) : _engine = engine,
       _toolContext = toolContext,
       _basicTools = basicTools ?? {},
       _extendedTools = extendedTools ?? {};

  final LlmEngine _engine;
  final ToolContext? _toolContext;
  final Map<String, AgentTool> _basicTools;
  final Map<String, ExtendedTool> _extendedTools;

  ConversationContext? _conversationContext;
  ToolPreferenceTracker? _preferenceTracker;

  LlmChatSession? _session;
  List<LlmToolSchema>? _cachedToolSchemas;

  bool _cancelled = false;

  final _riskClassifier = RiskClassifier();
  final _loopDetector = LoopDetector();
  final _confirmationGate = ConfirmationGate();
  final _permissionGate = PermissionGate();
  final _auditLog = AuditLog();
  late final ErrorRecovery _errorRecovery = ErrorRecovery(
    availableTools: _allToolNames,
  );

  PermissionChecker? _permissionChecker;

  Set<String> get _allToolNames => {
    ..._basicTools.keys,
    ..._extendedTools.keys,
  };

  @override
  void setPermissionChecker(PermissionChecker? checker) {
    _permissionChecker = checker;
  }

  LlmChatSession _ensureSession() {
    if (_session == null) {
      _session = _engine.createSession(_systemPrompt);
      _cachedToolSchemas = _buildToolSchemas();
    }
    return _session!;
  }

  static const _tag = 'AIOS-React';

  String get _systemPrompt {
    final base = StringBuffer()
      ..writeln('You are AIOS, an on-device phone assistant.')
      ..writeln(
        "Use tools to help the user. Respond concisely in the user's language.",
      )
      ..writeln(
        'When calling a tool, ALWAYS include ALL required parameters as JSON.',
      )
      ..writeln('Example: calculator needs {"expression": "2+3"}')
      ..writeln('If no tool is needed, answer directly.')
      ..writeln()
      ..writeln('CRITICAL RULES FOR screen_action:')
      ..writeln(
        '- Each screen_action call does exactly ONE action. '
        'Do NOT mix parameters from different actions in one call.',
      )
      ..writeln(
        '- NEVER put "content" and "global_action" in the same call. '
        'They belong to different actions.',
      )
      ..writeln('- Search workflow requires 3 SEPARATE calls:')
      ..writeln('  1) {"action":"tap","text":"검색"} → open search field')
      ..writeln('  2) {"action":"type","content":"query"} → type text')
      ..writeln('  3) {"action":"global","global_action":"enter"} → submit')
      ..writeln(
        '- Complex tasks always need MULTIPLE sequential tool calls. '
        'Do not stop until the task is fully done.',
      );

    final contextStr = _conversationContext?.toPromptContext() ?? '';
    if (contextStr.isNotEmpty) {
      base
        ..writeln()
        ..write(contextStr);
    }

    final prefStr = _preferenceTracker?.toPromptContext() ?? '';
    if (prefStr.isNotEmpty) {
      base
        ..writeln()
        ..write(prefStr);
    }

    return base.toString();
  }

  List<LlmToolSchema> _buildToolSchemas() {
    final schemas = <LlmToolSchema>[];
    for (final tool in _basicTools.values) {
      schemas.add(_toolToSchema(tool.name, tool.toolPrompt, tool.parameters));
    }
    for (final tool in _extendedTools.values) {
      schemas.add(_toolToSchema(tool.name, tool.toolPrompt, tool.parameters));
    }
    return schemas;
  }

  LlmToolSchema _toolToSchema(
    String name,
    String toolPrompt,
    String parameters,
  ) {
    return LlmToolSchema(
      name: name,
      description: toolPrompt,
      parameters: _parseParams(parameters),
    );
  }

  List<LlmToolParamSchema> _parseParams(String paramsStr) {
    try {
      final decoded = jsonDecode(paramsStr);
      if (decoded is! Map<String, dynamic>) return [];
      return decoded.entries.map((e) {
        final desc = e.value.toString();
        final type = _detectType(desc);
        if (desc.contains('|')) {
          final values = desc
              .split('|')
              .map((v) => v.replaceAll(RegExp('[^a-z_]'), ''))
              .where((v) => v.isNotEmpty)
              .toList();
          if (values.isNotEmpty) {
            return LlmToolParamSchema(
              name: e.key,
              description: desc,
              type: type,
              required: true,
              isEnum: true,
              enumValues: values,
            );
          }
        }
        final example = _extractExample(desc);
        return LlmToolParamSchema(
          name: e.key,
          description: desc,
          type: type,
          required: !_isOptional(desc),
          example: example,
        );
      }).toList();
    } on Object {
      return [];
    }
  }

  String _detectType(String desc) {
    final lower = desc.toLowerCase();
    if (lower.contains('boolean') || lower.contains('(true|false)')) {
      return 'boolean';
    }
    if (lower.contains('integer') || lower.contains('int ')) {
      return 'integer';
    }
    if (lower.contains('number') || lower.contains('float')) {
      return 'number';
    }
    return 'string';
  }

  bool _isOptional(String desc) {
    return desc.toLowerCase().contains('optional');
  }

  String? _extractExample(String desc) {
    final match = RegExp(
      r'\(e\.g\.?\s*([^)]+)\)',
      caseSensitive: false,
    ).firstMatch(desc);
    return match?.group(1)?.trim();
  }

  @override
  Future<AgentResult> execute(
    String prompt, {
    int maxIterations = 8,
    int maxTokens = 512,
    void Function(AgentStep)? onStep,
  }) async {
    _cancelled = false;
    _loopDetector.reset();
    _errorRecovery.reset();
    final steps = <AgentStep>[];

    print(
      '[$_tag] Agent run: '
      'prompt="${truncate(prompt, 50)}", '
      'maxIter=$maxIterations',
    );

    final runStartTime = DateTime.now();
    const maxRunDuration = Duration(seconds: 120);

    try {
      steps.add(AgentStep('thought', 'Processing: $prompt'));
      onStep?.call(steps.last);

      final session = _ensureSession();

      final tools = _cachedToolSchemas!;
      final generationConfig = LlmGenerationConfig(
        temperature: 0.7,
        topP: 0.9,
        maxTokens: maxTokens,
      );

      var userParts = [LlmContentPart.text(prompt)];

      for (var i = 0; i < maxIterations; i++) {
        if (_cancelled) break;

        final elapsed = DateTime.now().difference(runStartTime);
        if (elapsed > maxRunDuration) {
          steps.add(
            AgentStep('thought', 'Time limit reached (${elapsed.inSeconds}s).'),
          );
          onStep?.call(steps.last);
          break;
        }

        steps.add(AgentStep('thought', 'Thinking (step ${i + 1})...'));
        onStep?.call(steps.last);
        onStep?.call(const AgentStep('thinking_start', ''));

        String fullContent = '';
        final Map<int, _ToolCallAccumulator> toolCallBuilders = {};

        try {
          await for (final chunk in session.chat(
            userParts,
            config: generationConfig,
            tools: tools,
          )) {
            if (_cancelled) break;
            if (chunk.text != null) fullContent += chunk.text!;

            if (chunk.toolCallDeltas != null) {
              for (final tc in chunk.toolCallDeltas!) {
                toolCallBuilders.putIfAbsent(
                  tc.index,
                  _ToolCallAccumulator.new,
                );
                final builder = toolCallBuilders[tc.index]!;
                if (tc.id != null) builder.id = tc.id;
                if (tc.name != null) builder.name = tc.name;
                if (tc.arguments != null) {
                  builder.arguments += tc.arguments!;
                }
              }
            }
          }
        } on Object catch (e) {
          print('[$_tag] ERROR: generation stream error - $e');
        }

        onStep?.call(const AgentStep('thinking_end', ''));

        print(
          '[$_tag] Step $i: '
          'content="${truncate(fullContent, 200)}", '
          'toolCalls=${toolCallBuilders.length}',
        );

        if (toolCallBuilders.isEmpty) {
          final answer = fullContent.trim();
          if (answer.isNotEmpty) {
            steps.add(AgentStep('answer', answer));
            onStep?.call(steps.last);
            _recordTurn(prompt, steps);
            return AgentResult(steps: steps, success: true);
          }

          final retryCount = steps
              .where((s) => s.type == 'phase1_retry')
              .length;
          if (retryCount < 2) {
            steps.add(
              AgentStep('phase1_retry', 'Format retry ${retryCount + 1}/2'),
            );
            onStep?.call(steps.last);
            userParts = [
              const LlmContentPart.text(
                'Please use a tool or provide a direct answer.',
              ),
            ];
            continue;
          }

          steps.add(const AgentStep('answer', '요청을 처리하지 못했습니다.'));
          onStep?.call(steps.last);
          break;
        }

        for (final entry in toolCallBuilders.entries) {
          if (_cancelled) break;

          final builder = entry.value;
          final toolName = builder.name ?? '';
          print('[$_tag] Processing tool: $toolName');
          Map<String, dynamic> toolArgs = {};
          if (builder.arguments.isNotEmpty) {
            toolArgs = tryParseToolJson(builder.arguments, _tag);
          }
          print('[$_tag] Tool args: $toolArgs');

          if (toolArgs.isEmpty) {
            final inferred = _inferToolArgs(toolName, prompt);
            if (inferred != null) {
              toolArgs = inferred;
              print('[$_tag] Auto-inferred args for $toolName: $toolArgs');
            }
          }

          final argsJson = jsonEncode(toolArgs);

          steps.add(
            AgentStep(
              'action',
              'Using tool: $toolName',
              toolName: toolName,
              toolArgs: argsJson,
            ),
          );
          onStep?.call(steps.last);

          final requiredPerm = ToolPermissionMapper.getRequiredPermission(
            toolName,
            argsJson,
          );
          if (requiredPerm != null) {
            final hasPermission = await _checkPermission(requiredPerm.key);
            if (!hasPermission) {
              final granted = await _permissionGate.requestPermission(
                requiredPerm,
                toolName,
                onStep ?? (_) {},
              );
              if (!granted) {
                final result = 'Error: ${requiredPerm.displayName} 권한이 거부되었습니다';
                steps.add(
                  AgentStep(
                    'observation',
                    result,
                    toolName: toolName,
                    toolResult: result,
                  ),
                );
                onStep?.call(steps.last);
                session.addToolResult(toolName, result);
                continue;
              }
            }
          }

          final risk = _riskClassifier.classify(toolName, argsJson);
          print('[$_tag] Risk: $risk for $toolName');

          final validationError = await _validateTool(toolName, argsJson);
          if (validationError != null) {
            print('[$_tag] Validation error: $validationError');
            steps.add(
              AgentStep(
                'observation',
                validationError,
                toolName: toolName,
                toolResult: validationError,
              ),
            );
            onStep?.call(steps.last);
            session.addToolResult(toolName, validationError);
            continue;
          }

          if (risk == ToolRisk.high || risk == ToolRisk.critical) {
            print('[$_tag] ConfirmationGate: requesting for $toolName');
            final approved = await _confirmationGate.requestConfirmation(
              risk,
              toolName,
              argsJson,
              onStep ?? (_) {},
            );
            print('[$_tag] ConfirmationGate: approved=$approved');
            if (!approved) {
              _auditLog.add(toolName, argsJson, risk, false, 'Cancelled');
              const result = 'Action cancelled by user';
              steps.add(
                AgentStep(
                  'observation',
                  result,
                  toolName: toolName,
                  toolResult: result,
                ),
              );
              onStep?.call(steps.last);
              session.addToolResult(toolName, result);
              continue;
            }
          }

          print('[$_tag] Executing tool: $toolName');
          final toolResult = await _executeToolDirect(toolName, toolArgs);
          final toolContent = toolResult.toContent();
          print(
            '[$_tag] Tool result (${toolContent.length} chars): ${truncate(toolContent, 200)}',
          );

          _auditLog.add(toolName, argsJson, risk, true, toolContent);
          _preferenceTracker?.recordToolUse(toolName);

          steps.add(
            AgentStep(
              'observation',
              toolContent,
              toolName: toolName,
              toolResult: toolContent,
            ),
          );
          onStep?.call(steps.last);

          session.addToolResult(toolName, toolContent);

          final recoveryHint = _errorRecovery.analyze(
            toolName,
            argsJson,
            toolResult,
          );
          if (recoveryHint != null && recoveryHint.promptNudge.isNotEmpty) {
            print('[$_tag] Recovery: type=${recoveryHint.type}');
            if (recoveryHint.shouldRetry) {
              final toolPrompt = _getToolPrompt(toolName);
              userParts = [
                LlmContentPart.text(
                  '${recoveryHint.promptNudge}\n'
                  '$toolPrompt',
                ),
              ];
            }
          }

          final loopResult = _loopDetector.record(
            toolName,
            argsJson,
            toolResult,
          );
          if (loopResult is LoopForceBreak) {
            steps.add(const AgentStep('answer', '작업이 반복 감지로 중단되었습니다.'));
            onStep?.call(steps.last);
            _recordTurn(prompt, steps);
            return AgentResult(steps: steps, success: false);
          }
        }

        userParts = [];
      }

      if (steps.every((s) => s.type != 'answer')) {
        steps.add(const AgentStep('answer', '작업을 완료하지 못했습니다.'));
        onStep?.call(steps.last);
      }
    } on Object catch (e) {
      print('[$_tag] ERROR: Agent run crashed: $e');
      if (steps.every((s) => s.type != 'answer')) {
        steps.add(AgentStep('answer', '오류가 발생했습니다: $e'));
        onStep?.call(steps.last);
      }
    }

    final success = steps.any((s) => s.type == 'answer');
    _recordTurn(prompt, steps);
    return AgentResult(steps: steps, success: success);
  }

  Future<bool> _checkPermission(String permissionKey) async {
    if (_permissionChecker == null) return true;
    try {
      return await _permissionChecker!(permissionKey);
    } on Object catch (e) {
      print('[$_tag] WARN: permission check failed - $e');
      return true;
    }
  }

  Future<String?> _validateTool(String name, String argsJson) async {
    final basicTool = _basicTools[name];
    if (basicTool != null) {
      return basicTool.validate(argsJson);
    }

    final extendedTool = _extendedTools[name];
    if (extendedTool != null) {
      final ctx = _toolContext;
      if (ctx == null) {
        return 'Error: ToolContext not initialized';
      }
      return extendedTool.validate(argsJson, ctx);
    }

    return null;
  }

  Future<ToolResult> _executeToolDirect(
    String name,
    Map<String, dynamic> args,
  ) async {
    final argsJson = jsonEncode(args);
    final basicTool = _basicTools[name];
    final extendedTool = _extendedTools[name];

    if (basicTool != null) {
      return basicTool.execute(argsJson);
    }

    if (extendedTool != null) {
      final ctx = _toolContext;
      if (ctx == null) {
        return const ToolResult.err('ToolContext not initialized');
      }
      return extendedTool.execute(argsJson, ctx);
    }

    return ToolResult.err("Unknown tool '$name'");
  }

  String _getToolPrompt(String name) {
    final basicTool = _basicTools[name];
    if (basicTool != null) {
      return 'Tool "$name" help: ${basicTool.toolPrompt}';
    }
    final extendedTool = _extendedTools[name];
    if (extendedTool != null) {
      return 'Tool "$name" help: ${extendedTool.toolPrompt}';
    }
    return '';
  }

  Map<String, dynamic>? _inferToolArgs(String toolName, String userMessage) {
    return inferToolArgs(toolName, userMessage);
  }

  void _recordTurn(String userMessage, List<AgentStep> steps) {
    if (_conversationContext == null) return;
    final answerStep = steps.where((s) => s.type == 'answer').lastOrNull;
    if (answerStep == null) return;
    final toolSteps = steps
        .where((s) => s.type == 'action' && s.toolName.isNotEmpty)
        .toList();
    final toolUsed = toolSteps.isNotEmpty ? toolSteps.first.toolName : null;
    _conversationContext!.addTurn(
      userMessage,
      answerStep.content,
      toolUsed: toolUsed,
    );
  }

  @override
  void cancel() {
    _cancelled = true;
    _engine.cancelGeneration();
    _confirmationGate.cancel();
    _permissionGate.cancel();
  }

  @override
  void resolveConfirmation(bool approved) {
    _confirmationGate.resolve(approved);
  }

  @override
  void resolvePermission(bool granted) {
    _permissionGate.resolve(granted);
  }

  @override
  String getToolManifest() {
    final lines = <String>[];
    for (final tool in _basicTools.values) {
      lines.add('- ${tool.name}: ${tool.description}');
    }
    for (final tool in _extendedTools.values) {
      lines.add('- ${tool.name}: ${tool.description}');
    }
    return lines.join('\n');
  }

  @override
  List<({String role, String content})> getConversationHistory() => [];

  @override
  void clearHistory() {
    _session = null;
    _cachedToolSchemas = null;
  }

  @override
  Future<void> warmup() async {
    await _engine.warmup();
  }

  @override
  void setConversationContext(ConversationContext? context) {
    _conversationContext = context;
  }

  @override
  void setToolPreferenceTracker(ToolPreferenceTracker? tracker) {
    _preferenceTracker = tracker;
  }
}

class _ToolCallAccumulator {
  String? id;
  String? name;
  String arguments = '';
}

File: lib/src/agent/react-strategy.ts
import type { AgentStep, AgentResult, ToolResult } from '../types/agent';
import type { LlmToolSchema, LlmProviderConfig, LlmGenerationConfig } from '../llm/types';
import { LlmRemoteSession } from '../llm/session';
import { OpenAiClient } from '../llm/openai-client';
import { RiskClassifier } from './risk-classifier';
import { ErrorRecovery } from './error-recovery';
import { LoopDetector } from './loop-detector';
import { ConversationContext } from './conversation-context';
import { ToolPreferenceTracker } from './tool-preference-tracker';
import { tryParseToolJson } from './tool-json-parser';
import { inferToolArgs } from './tool-arg-inference';
import { truncate } from './truncate';
import { toContent, toolResultErr } from '../types/agent';
import type { AgentTool } from '../tools/types';

const TAG = 'AIOS-React';

export class ReactStrategy {
  private session: LlmRemoteSession | null = null;
  private cachedToolSchemas: LlmToolSchema[] | null = null;
  private cancelled = false;
  private client: OpenAiClient | null = null;

  private riskClassifier = new RiskClassifier();
  private loopDetector = new LoopDetector();
  private errorRecovery = new ErrorRecovery();
  private conversationContext: ConversationContext | null = null;
  private preferenceTracker: ToolPreferenceTracker | null = null;
  private resolveConfirmationFn: ((approved: boolean) => void) | null = null;
  private generationConfig: LlmGenerationConfig = { temperature: 1.0, topP: 0.95, maxTokens: 8192 };

  constructor(
    private readonly agentTools: Map<string, AgentTool>,
    private readonly config: LlmProviderConfig,
  ) {}

  setConversationContext(ctx: ConversationContext | null): void {
    this.conversationContext = ctx;
  }

  setToolPreferenceTracker(tracker: ToolPreferenceTracker | null): void {
    this.preferenceTracker = tracker;
  }

  setGenerationConfig(config: Partial<LlmGenerationConfig>): void {
    this.generationConfig = { ...this.generationConfig, ...config };
  }

  private ensureSession(): LlmRemoteSession {
    if (!this.session) {
      this.client = new OpenAiClient(this.config);
      this.session = new LlmRemoteSession(this.client, this.systemPrompt);
      this.cachedToolSchemas = this.buildToolSchemas();
    }
    return this.session;
  }

  private get systemPrompt(): string {
    const lines: string[] = [
      'You are AIOS, an on-device phone assistant.',
      "Use tools to help the user. Respond concisely in the user's language.",
      'When calling a tool, ALWAYS include ALL required parameters as JSON.',
      'If no tool is needed, answer directly.',
      '',
      'Available tools:',
    ];
    for (const tool of this.agentTools.values()) {
      lines.push(`- ${tool.name}: ${tool.description}`);
    }
    const ctx = this.conversationContext?.toPromptContext() ?? '';
    if (ctx) lines.push('', ctx);
    const pref = this.preferenceTracker?.toPromptContext() ?? '';
    if (pref) lines.push('', pref);
    return lines.join('\n');
  }

  private buildToolSchemas(): LlmToolSchema[] {
    const schemas: LlmToolSchema[] = [];
    for (const tool of this.agentTools.values()) {
      schemas.push({
        name: tool.name,
        description: tool.toolPrompt,
        parameters: this.parseParams(tool.parameters),
      });
    }
    return schemas;
  }

  private parseParams(paramsStr: string): LlmToolSchema['parameters'] {
    try {
      const decoded = JSON.parse(paramsStr);
      if (typeof decoded !== 'object' || decoded === null) return [];
      return Object.entries(decoded).map(([key, desc]) => ({
        name: key,
        description: String(desc),
        type: 'string' as const,
        required: !String(desc).toLowerCase().includes('optional'),
      }));
    } catch {
      return [];
    }
  }

  async execute(
    prompt: string,
    onStep?: (step: AgentStep) => void,
    maxIterations = 8,
  ): Promise<AgentResult> {
    this.cancelled = false;
    this.loopDetector.reset();
    this.errorRecovery.reset();
    const steps: AgentStep[] = [];

    console.log(`[${TAG}] Agent run: prompt="${truncate(prompt, 50)}"`);
    const runStartTime = Date.now();
    const maxRunDuration = 120_000;

    try {
      steps.push({ type: 'thought', content: `Processing: ${prompt}` });
      onStep?.(steps[steps.length - 1]);

      const session = this.ensureSession();
      const tools = this.cachedToolSchemas!;
      const config = this.generationConfig;
      let userParts = [prompt];
      let continuationCount = 0;
      const MAX_CONTINUATIONS = 3;

      for (let i = 0; i < maxIterations; i++) {
        if (this.cancelled) break;
        if (Date.now() - runStartTime > maxRunDuration) break;

        steps.push({ type: 'thought', content: `Thinking (step ${i + 1})...` });
        onStep?.(steps[steps.length - 1]);

        let fullContent = '';
        const toolCallBuilders = new Map<number, { id: string; name: string; arguments: string }>();
        const seenToolNames = new Set<number>();
        let lastStreamEmit = 0;
        let lastFinishReason = '';
        const STREAM_THROTTLE_MS = 50;

        try {
          for await (const chunk of session.chat(userParts, config, tools)) {
            if (this.cancelled) break;
            if (chunk.finishReason) lastFinishReason = chunk.finishReason;
            if (chunk.text) {
              fullContent += chunk.text;
              const now = Date.now();
              if (now - lastStreamEmit > STREAM_THROTTLE_MS) {
                lastStreamEmit = now;
                onStep?.({ type: 'streaming_text', content: fullContent });
              }
            }
            if (chunk.toolCallDeltas) {
              for (const tc of chunk.toolCallDeltas) {
                const b = toolCallBuilders.get(tc.index) || { id: '', name: '', arguments: '' };
                if (tc.id) b.id = tc.id;
                if (tc.name) b.name = tc.name;
                if (tc.arguments) b.arguments += tc.arguments;
                toolCallBuilders.set(tc.index, b);
                if (tc.name && !seenToolNames.has(tc.index)) {
                  seenToolNames.add(tc.index);
                  onStep?.({ type: 'tool_call_start', content: `${tc.name} 준비 중...`, toolName: tc.name });
                }
              }
            }
          }
        } catch (e) {
          console.error(`[${TAG}] Stream error:`, e);
        }

        if (fullContent) {
          onStep?.({ type: 'streaming_text', content: fullContent });
        }

        if (toolCallBuilders.size === 0) {
          if (lastFinishReason === 'length' && fullContent && continuationCount < MAX_CONTINUATIONS) {
            continuationCount++;
            steps.push({ type: 'continuation', content: `계속 생성 중... (${continuationCount}/${MAX_CONTINUATIONS})` });
            onStep?.(steps[steps.length - 1]);
            userParts = ['이어서 계속 작성하세요.'];
            continue;
          }

          const answer = fullContent.trim();
          if (answer) {
            steps.push({ type: 'answer', content: answer });
            onStep?.(steps[steps.length - 1]);
            this.recordTurn(prompt, steps);
            return { steps, success: true };
          }
          const retryCount = steps.filter((s) => s.type === 'phase1_retry').length;
          if (retryCount < 2) {
            steps.push({ type: 'phase1_retry', content: `Retry ${retryCount + 1}/2` });
            onStep?.(steps[steps.length - 1]);
            userParts = ['Please use a tool or provide a direct answer.'];
            continue;
          }
          steps.push({ type: 'answer', content: '요청을 처리하지 못했습니다.' });
          onStep?.(steps[steps.length - 1]);
          break;
        }

        for (const [, builder] of toolCallBuilders) {
          if (this.cancelled) break;
          const toolName = builder.name;
          let toolArgs = tryParseToolJson(builder.arguments);
          if (Object.keys(toolArgs).length === 0) {
            const inferred = inferToolArgs(toolName, prompt);
            if (inferred) toolArgs = inferred;
          }
          const argsJson = JSON.stringify(toolArgs);
          steps.push({ type: 'action', content: `Using tool: ${toolName}`, toolName, toolArgs: argsJson });
          onStep?.(steps[steps.length - 1]);

          const risk = this.riskClassifier.classify(toolName, argsJson);
          const validationError = await this.agentTools.get(toolName)?.validate?.(argsJson);
          if (validationError) {
            steps.push({ type: 'observation', content: validationError, toolName, toolResult: validationError });
            onStep?.(steps[steps.length - 1]);
            session.addToolResult(toolName, validationError);
            continue;
          }

          if (risk === 'high' || risk === 'critical') {
            steps.push({ type: 'confirmation_required', content: `Confirm: ${toolName}`, toolName, toolArgs: argsJson, riskLevel: risk });
            onStep?.(steps[steps.length - 1]);
            const approved = await this.waitForConfirmation();
            if (!approved) {
              const msg = 'Action cancelled by user';
              steps.push({ type: 'observation', content: msg, toolName, toolResult: msg });
              onStep?.(steps[steps.length - 1]);
              session.addToolResult(toolName, msg);
              continue;
            }
          }

          const toolResult = await this.executeTool(toolName, argsJson);
          const toolContent = toContent(toolResult);
          this.preferenceTracker?.recordToolUse(toolName);
          steps.push({ type: 'observation', content: toolContent, toolName, toolResult: toolContent });
          onStep?.(steps[steps.length - 1]);
          session.addToolResult(toolName, toolContent);

          this.errorRecovery.analyze(toolName, argsJson, toolResult);

          const loopResult = this.loopDetector.record(toolName, argsJson, toolResult);
          if (loopResult.type === 'forceBreak') {
            steps.push({ type: 'answer', content: '작업이 반복 감지로 중단되었습니다.' });
            onStep?.(steps[steps.length - 1]);
            this.recordTurn(prompt, steps);
            return { steps, success: false };
          }
        }
        userParts = [];
      }

      if (!steps.some((s) => s.type === 'answer')) {
        steps.push({ type: 'answer', content: '작업을 완료하지 못했습니다.' });
        onStep?.(steps[steps.length - 1]);
      }
    } catch (e) {
      console.error(`[${TAG}] Crashed:`, e);
      if (!steps.some((s) => s.type === 'answer')) {
        steps.push({ type: 'answer', content: `오류: ${e}` });
        onStep?.(steps[steps.length - 1]);
      }
    }

    this.recordTurn(prompt, steps);
    return { steps, success: steps.some((s) => s.type === 'answer') };
  }

  private async executeTool(name: string, argsJson: string): Promise<ToolResult> {
    const tool = this.agentTools.get(name);
    if (!tool) return toolResultErr(`Unknown tool '${name}'`);
    try {
      return await tool.execute(argsJson);
    } catch (e) {
      return toolResultErr(`${e}`);
    }
  }

  private recordTurn(userMessage: string, steps: AgentStep[]): void {
    if (!this.conversationContext) return;
    const answerStep = [...steps].reverse().find((s) => s.type === 'answer');
    if (!answerStep) return;
    const actionStep = steps.find((s) => s.type === 'action' && s.toolName);
    this.conversationContext.addTurn(userMessage, answerStep.content, actionStep?.toolName);
  }

  cancel(): void {
    this.cancelled = true;
    this.client?.cancel();
    if (this.resolveConfirmationFn) {
      this.resolveConfirmationFn(false);
      this.resolveConfirmationFn = null;
    }
  }

  resolveConfirmation(approved: boolean): void {
    this.resolveConfirmationFn?.(approved);
    this.resolveConfirmationFn = null;
  }

  private waitForConfirmation(): Promise<boolean> {
    return new Promise((resolve) => {
      this.resolveConfirmationFn = resolve;
      setTimeout(() => {
        if (this.resolveConfirmationFn === resolve) {
          this.resolveConfirmationFn = null;
          resolve(false);
        }
      }, 60_000);
    });
  }

  clearHistory(): void {
    this.session = null;
    this.cachedToolSchemas = null;
  }
}
